openmmapi/include/openmm/NoseHooverIntegrator.h

```diff
@@ -0,0 +1,245 @@
+ * and/or sell copies of the Software, and to permit persons to whom the      *
+ * Software is furnished to do so, subject to the following conditions:       *
+ *                                                                            *
+ * The above copyright notice and this permission notice shall be included in *
+ * all copies or substantial portions of the Software.                        *
+ *                                                                            *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR *
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,   *
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL    *
+ * THE AUTHORS, CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,    *
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR      *
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE  *
+ * USE OR OTHER DEALINGS IN THE SOFTWARE.                                     *
+ * -------------------------------------------------------------------------- */
+
+#include "Integrator.h"
+#include "openmm/State.h"
+#include "openmm/Kernel.h"
+#include "openmm/NoseHooverChain.h"
+#include "internal/windowsExport.h"
+
+namespace OpenMM {
+
+class System;
+
+/**
+ * This is an Integrator which simulates a System using one or more Nose Hoover chain
+ * thermostats, using the velocity Verlet propagation algorithm.
+ */
+
+class OPENMM_EXPORT NoseHooverIntegrator : public Integrator {
+public:
+    /**
+     * Create a NoseHooverIntegrator.  This version creates a bare velocity Verlet integrator
+     * with no thermostats; any thermostats should be added by calling addThermostat.
+     * 
+     * @param stepSize the step size with which to integrate the system (in picoseconds)
+     */
+    explicit NoseHooverIntegrator(double stepSize);
+    /**
+     * Create a NoseHooverIntegrator.
+     *
+     * @param stepSize the step size with which to integrate the system (in picoseconds)
+     * @param system the system to be thermostated.  Note: this must be setup, i.e. all
+     *        particles should have been added, before calling this function.
+     * @param temperature the target temperature for the system.
+     * @param collisionFrequency the frequency of the interaction with the heat bath (in 1/ps).
+     * @param chainLength the number of beads in the Nose-Hoover chain.
+     * @param numMTS the number of step in the  multiple time step chain propagation algorithm.
+     * @param numYoshidaSuzuki the number of terms in the Yoshida-Suzuki multi time step decomposition
+     *        used in the chain propagation algorithm (must be 1, 3, or 5).
+     */
+    explicit NoseHooverIntegrator(double stepSize, System &system, double temperature, int collisionFrequnency,
+                                  int chainLength = 3, int numMTS = 3, int numYoshidaSuzuki = 3);
```

Comment:
Let's make the constructor arguments consistent with LangevinIntegrator:

```c++
NoseHooverIntegrator(double temperature, double collisionFrequnency, double stepSize, int chainLength=3, int numMTS=3, int numYoshidaSuzuki=3)
```

Having similar arguments but in a different order will lead to lots of confusion.

Requiring the System as an argument breaks some design assumptions and will lead to lots of problems.  Users are supposed to be able to create the System and Integrator independently of each other.  Just store whatever information gets passed to the constructor, and then you'll get a reference to the System when `initialize()` is called.